src/installer-generator/installer-generator.c

```diff
@@ -0,0 +1,156 @@
+} arg_enabled = INSTALLER_NO;
+static bool arg_disable_repart = true;
+static VolatileMode arg_volatile = VOLATILE_YES;
+
+static int parse_cmdline(const char *key, const char *value, void *data) {
+        int r;
+
+        assert(key);
+
+        if (proc_cmdline_key_streq(key, "systemd.installer")) {
+                if (!value)
+                        arg_enabled = INSTALLER_YES;
+                else if (streq(value, "auto"))
+                        arg_enabled = INSTALLER_AUTO;
+                else {
+                        r = parse_boolean(value);
+                        if (r < 0)
+                                log_warning_errno(r, "Failed to parse systemd.installer value \"%s\", ignoring: %m", value);
+                        arg_enabled = r ? INSTALLER_YES : INSTALLER_NO;
+                }
+        } else if (proc_cmdline_key_streq(key, "systemd.installer_disable_repart")) {
+                r = value ? parse_boolean(value) : true;
+                if (r < 0)
+                        log_warning_errno(r, "Failed to parse systemd.installer_disable_repart value \"%s\", ignoring: %m", value);
+                else
+                        arg_disable_repart = r;
+        } else if (proc_cmdline_key_streq(key, "systemd.installer_volatile")) {
+                if (!value)
+                        arg_volatile = VOLATILE_YES;
+                else if (streq(value, "overlay"))
+                        arg_volatile = VOLATILE_OVERLAY;
+                else {
+                        r = parse_boolean(value);
+                        if (r < 0)
+                                log_warning_errno(r, "Failed to parse systemd.installer_volatile value \"%s\", ignoring: %m", value);
+                        else
+                                arg_volatile = r ? VOLATILE_YES : VOLATILE_NO;
+                }
+        }
+
+        return 0;
+}
+
+static int conditionalize_unit(const char *unit) {
+        return write_drop_in(arg_dest, unit, 50, "installer-auto",
+                             "# Automatically generated by systemd-installer-generator\n\n"
+                             "[Service]\n"
+                             "Environment=SYSTEMD_INSTALLER_DISABLE_IF_REMOVABLE=1\n");
+}
+
+static int symlink_unit(const char *unit, const char *target) {
+        _cleanup_free_ char *p = NULL;
+
+        p = path_join(arg_dest, unit);
+        if (!p)
+                return log_oom();
+
+        if (symlink(target, p) < 0)
+                return log_error_errno(errno, "Failed to link unit %s -> %s: %m", unit, target);
+
+        return 0;
+}
+
+static int root_is_removable(void) {
+        dev_t devno;
+        _cleanup_(sd_device_unrefp) sd_device *dev = NULL;
+        int r;
+
+        r = blockdev_get_root(LOG_ERR, &devno);
+        if (r < 0)
+                return r;
+        if (r == 0)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Root file system not backed by a (single) whole block device.");
+
```

Comment:
event hough device_is_removable() goes up the tree anyway I think it would make sense to call block_get_whole_disk() explicitly first here.